### PR TITLE
buffer_filter: allow to set the maximum size Without buffering data immediately

### LINKED ISSUE #32205: Allow configuring decoder buffer limit per Route via filter
*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Setting the per Route decoder buffer limit would be useful for increasing the limit in some routes that need to deal with big request body in a fully buffering filter. Currently, to configure a per Route level decoder buffer limit, we can use the `buffer` filter. However, there is a side effect that the `buffer` filter will buffer all the request body. In some cases, the fully buffering is optional, for example, a WAF filter may skip the request body according to the configuration. Buffering the whole request body early will cause an unexpected 413 response.

What we actually need is just calling the `setDecoderBufferLimit` like
https://github.com/envoyproxy/envoy/blob/62e7c593748b2ceb3b80a3398e0a937a92edf95d/source/extensions/filters/http/buffer/buffer_filter.cc#L63

Can we add a new option `set_limit_only` in the `buffer` filter to call the `setDecoderBufferLimit` without doing the buffering? Or maybe we need to add a new filter for this purpose? I will submit a PR if this feature is accepted.

### PR DESCRIPTION
**Pull request by spacewander**
Commit Message: buffer_filter: allow to set the maximum size Without buffering data immediately
Additional Description:
Risk Level: Low, doesn't change the default behavior
Testing: Unit & Integration
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #32205
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by spacewander** [comment]
I am blocked by how to run the corresponding test. In the integration test, I used a Lua filter to simulate a filter that buffers the request. However, I get the error `C++ exception with the description "Unable to parse JSON as proto (INVALID_ARGUMENT: could not find @type 'type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua'):  ...`.

I have checked that the type of Lua filter is correct. I don't know why this type is not found.

Can anyone help me, please? Thanks!

File: api/envoy/extensions/filters/http/buffer/v3/buffer.proto [hunk 1]
```diff
@@ -28,6 +28,11 @@ message Buffer {
   // manager will stop buffering and return a 413 response.
   google.protobuf.UInt32Value max_request_bytes = 1
       [(validate.rules).uint32 = {gt: 0}, (validate.rules).message = {required: true}];
+
+  // Set the maximum request size but doesn't buffer the data immediately.
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/buffer/v3/buffer.proto:32]: should "immediately" be replaced with "doesn't buffer the data in the Buffer filter"?

File: api/envoy/extensions/filters/http/buffer/v3/buffer.proto [hunk 2]
```diff
@@ -28,6 +28,11 @@ message Buffer {
...
+  // It's useful when the following filters may do the buffer themselves conditionally, so
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/buffer/v3/buffer.proto:33]: nit: perhaps this reads a bit clearer:

```
It's useful when subsequent filters may do the buffering themselves conditionally, so the Buffer filter doesn't buffer prematurely.
```

File: api/envoy/extensions/filters/http/buffer/v3/buffer.proto [hunk 3]
```diff
@@ -28,6 +28,11 @@ message Buffer {
...
+  // we won't buffer the data too early.
+  bool set_limit_only = 3;
```
Comment (abeyad) [inline api/envoy/extensions/filters/http/buffer/v3/buffer.proto:35]: What happens if `max_request_bytes` isn't set, is this field ignored?  If so, a comment to that effect would be great.

**Review by abeyad — COMMENTED** [review]
Thanks!  Left some API (proto change) comments.

File: api/envoy/extensions/filters/http/buffer/v3/buffer.proto (hunk 3 above)
Comment (spacewander) [reply api/envoy/extensions/filters/http/buffer/v3/buffer.proto:35]: > What happens if `max_request_bytes` isn't set, is this field ignored? If so, a comment to that effect would be great.

https://github.com/envoyproxy/envoy/blob/372a262894f028431cb84f3a1d361fb8fb6a2518/api/envoy/extensions/filters/http/buffer/v3/buffer.proto#L29-L30

The `max_request_bytes` is a required field, so it must be set.

**Comment by spacewander** [comment]
@abeyad 
Thanks for your suggestion! Updated.

**Review by abeyad — APPROVED** [review]
/lgtm api

**Comment by spacewander** [comment]
> I am blocked by how to run the corresponding test. In the integration test, I used a Lua filter to simulate a filter that buffers the request. However, I get the error `C++ exception with the description "Unable to parse JSON as proto (INVALID_ARGUMENT: could not find @type 'type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua'): ...`.
> 
> I have checked that the type of Lua filter is correct. I don't know why this type is not found.
> 
> Can anyone help me, please? Thanks!

@abeyad 
I have an issue with writing a test to simulate the case that a subsequence filter does the buffering. Would you help me please?

**Comment by adisuissa** [comment]
> > I am blocked by how to run the corresponding test. In the integration test, I used a Lua filter to simulate a filter that buffers the request. However, I get the error `C++ exception with the description "Unable to parse JSON as proto (INVALID_ARGUMENT: could not find @type 'type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua'): ...`.
> > I have checked that the type of Lua filter is correct. I don't know why this type is not found.
> > Can anyone help me, please? Thanks!
> 
> @abeyad I have an issue with writing a test to simulate the case that a subsequence filter does the buffering. Would you help me please?

This is probably because the filter isn't linked.
Will it be possible to use one of the [integration tests filters](https://github.com/envoyproxy/envoy/tree/main/test/integration/filters) instead? This will reduce coupling with other filters.
Note that adding the filter to the build target (in the BUILD file) is also needed.

**Comment by spacewander** [comment]
@adisuissa 
Thanks for your suggestion! I will update the test soon.

**Review by alyssawilk — COMMENTED** [review]
sorry, is the goal of this PR simply to have a way of calling setDecoderBufferLimit on the filter manager?  I don't think it makes sense to add config to the buffer filter for this. Is there a reason  you can't use the existing flow control knobs?
https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control
/wait

**Comment by spacewander** [comment]
> sorry, is the goal of this PR simply to have a way of calling setDecoderBufferLimit on the filter manager? I don't think it makes sense to add config to the buffer filter for this. Is there a reason you can't use the existing flow control knobs? https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control /wait

There is a user case that people may want to increase the limit for a separate route. For example, a route that is covered by a WAF rule may need to increase its limit so no 413 response is triggered unexpectedly. As discussed in https://github.com/envoyproxy/envoy/issues/32205, there is no solution to cover this user case. Using the buffer filter is OK, but it will buffer the whole request unconditionally.

We can also add this feature to a new HTTP filter. One thing I am concerned about is whether adding a new filter for this purpose is overkill. But it is up to you to make the decision.

**Comment by alyssawilk** [comment]
We already have envoy per-route config
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto
I think changing buffer limits makes more sense there than in a filter many may not use.

**Comment by spacewander** [comment]
> We already have envoy per-route config https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto I think changing buffer limits makes more sense there than in a filter many may not use.

Do you mean adding a `max_request_bytes` field to RouteConfiguration message?

**Comment by alyssawilk** [comment]
something like that yes.  cc @envoyproxy/api-shepherds

**Comment by spacewander** [comment]
Will submit a new one with the new method.

**Pull request closed without merging**